src/foam/nanos/crunch/JavaCrunchService.java

```diff
@@ -0,0 +1,140 @@
+package foam.nanos.crunch;
+
+import foam.core.X;
+import foam.dao.ArraySink;
+import foam.dao.DAO;
+import foam.mlang.predicate.Predicate;
+import foam.nanos.auth.Subject;
+import foam.nanos.auth.User;
+import foam.nanos.logger.Logger;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import static foam.mlang.MLang.*;
+
+public class JavaCrunchService implements CrunchService {
+  public List getGrantPath(X x, String rootId) {
+    Logger logger = (Logger) x.get("logger");
+
+    DAO prerequisiteDAO = (DAO) x.get("prerequisiteCapabilityJunctionDAO");
+    DAO capabilityDAO = (DAO) x.get("capabilityDAO");
+
+    // Lookup for indices of previously observed capabilities
+    Map<String,Integer> alreadyListed = new HashMap<String,Integer>();
+    // Return list (list is reversed at the end to put prerequisites first)
+    List grantPath = new ArrayList<>();
+
+    Queue<String> nextSources = new ArrayDeque<String>();
+    nextSources.add(rootId);
+
+    // Doing this instead of "this" could prevent unexpected behaviour
+    // in the incident CrunchService's getJunction method is decorated
+    CrunchService crunchService = (CrunchService) x.get("crunchService");
+
+    while ( nextSources.size() > 0 ) {
+      String sourceId = nextSources.poll();
+
+      UserCapabilityJunction ucj = crunchService.getJunction(x, sourceId);
+      if ( ucj != null && ucj.getStatus() == CapabilityJunctionStatus.GRANTED ) {
+        continue;
+      }
+
+      // Remove previously added prerequisite if one matches
+      if ( alreadyListed.containsKey(sourceId) ) {
+        int previousIndex = alreadyListed.get(sourceId);
+        grantPath.remove(previousIndex);
+
+        // Remove previously stored index of capability
+        alreadyListed.remove(sourceId);
+
+        // Shift remembered indexes, now that grantList has been shifted
+        Map<String,Integer> newAlreadyListed = new HashMap<String,Integer>();
+        for ( Map.Entry<String,Integer> entry : alreadyListed.entrySet() ) {
+          int newIndex = entry.getValue();
+          if ( newIndex > previousIndex ) newIndex--;
+        }
+        alreadyListed = newAlreadyListed;
```

Comment:
Assign alreadylisted to new hashmap? is this intended?

Reply:
This is intended... but I think I missed a step in that for loop

Reply:
resolved